### PR TITLE
[JENKINS-36871] JNLP4-connect implementation

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-36871](https://issues.jenkins-ci.org/browse/JENKINS-36871)

This adds the protocol as opt-in, however integrating the protocol required reworking some other classes.

`JnlpAgentReceiver` was actually broken (both in binary compatibility and functional compatibility) by the introduction of `JNLP3-connect` and we cannot identify any consumers other than a CloudBees closed source plugin, so breaking binary compatibility again is not critical

Todo
- [x] Test old remoting.jar against new code
- [x] Test new remoting.jar against old code
- [x] Restore the cookie behaviour (but done right this time)
- [x] ~~Perhaps investigate issuing clients with TLS certificates (but would require a UI for managing them)~~

@jenkinsci/code-reviewers @reviewbybees

See also https://github.com/jenkinsci/remoting/pull/92

### CODE REVIEW

**Comment by stephenc** [comment]
Needs matching release of remoting, but functionally complete

File: core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol.java [hunk 1]
```diff
@@ -56,10 +47,29 @@
  * @author Kohsuke Kawaguchi
  * @since 1.467
  */
-@Extension @Symbol("jnlp")
+@Extension
+@Symbol("jnlp")
```
Comment (jtnord) [inline core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol.java]: sneaky :-)

File: core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java [hunk 2]
```diff
@@ -155,7 +92,7 @@ protected String getNodeSecret(String nodeName) throws Failure {
             ENABLED = forceEnabled;
         else {
             byte hash = Util.fromHexString(Jenkins.getActiveInstance().getLegacyInstanceId())[0];
-            ENABLED = (hash%10)==0;
+            ENABLED = (hash % 10) == 0;
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java]: Causes merge conflict

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers Before going forward, it would be definitely great to know your opinion regarding this binary-incompatible change. @stephenc asked in the Dev List in Twitter and got no response, but it's not a 100% guarantee it's not being used anywhere. But actually it's highly unlikely

File: pom.xml [hunk 3]
```diff
@@ -180,7 +180,7 @@ THE SOFTWARE.
       <dependency>
         <groupId>org.jenkins-ci.main</groupId>
         <artifactId>remoting</artifactId>
-        <version>2.61</version>
+        <version>3.0-20160811.164749-2</version> <!-- TODO pick up release of remoting -->
```
Comment (oleg-nenashev) [inline pom.xml]: remoting-3 is ready

**Review by oleg-nenashev — COMMENTED** [review]
Remoting 3 is available. This PR is unblocked

File: pom.xml (hunk 3 above)
Comment (stephenc) [inline pom.xml]: Ok, I'll refresh this PR

**Comment by oleg-nenashev** [comment]
🐝 Test failure is unrelated

@jenkinsci/code-reviewers 
As per voting results from the previous governance meeting ([summary](http://meetings.jenkins-ci.org/jenkins-meeting/2016/jenkins-meeting.2016-10-12-18.00.html)), we can proceed with the common PR integration approach. 

Integration of remoting 3 is being discussed [here](https://groups.google.com/forum/#!searchin/jenkinsci-dev/remoting$203%7Csort:relevance/jenkinsci-dev/SS6z-09dHIs/Qv_N5KI1BQAJ). By now there are 3 +1s and no -1s. So technically this PR qualifies the merge criteria.

**Comment by stephenc** [comment]
w00t!

**Comment by jtnord** [comment]
How did I not see that this broke binary and source backwards compatibility...
Are we getting Jenkins 3.0 now :-)

**Comment by stephenc** [comment]
@jtnord... what ya complaining about... you're getting to rip out all the code that we open sourced and replace with references to the open sourced version

**Pull request merged**